src/emmy/differential.cljc

```diff
@@ -746,284 +539,117 @@
-       (->> (bare-terms dx)
-            (filterv sans-tag?)
-            (terms->differential)))
-     dx)))
-
-(defn tangent-part
-  "Returns a [[Differential]] containing only the terms of `dx` that contain the
-  supplied `tag`, i.e., the tangent component of `dx` with respect to `tag`.
-
-  If no tag is supplied, defaults to `([[max-order-tag]] dx)`.
-
-  NOTE: Every [[Differential]] can be factored into a dual number of the form
-
-      primal + (tangent * tag)
-
-  For each tag in any of its terms. [[tangent-part]] returns a [[Differential]]
-  representing `(tangent * tag)`, or 0 if `dx` contains no terms with the
-  supplied `tag`.
-
-  NOTE: the 2-arity case is similar to `([[extract-tangent]] dx tag)`; the only
-  difference is that [[extract-tangent]] drops the `dx` tag from all terms in
-  the returned value. Call [[extract-tangent]] if you want to drop `tag`."
-  ([dx] (tangent-part dx (max-order-tag dx)))
-  ([dx tag]
-   (if (differential? dx)
-     (->> (bare-terms dx)
-          (filterv #(tag-in-term? % tag))
-          (terms->differential))
-     0)))
-
-(defn primal-tangent-pair
-  "Returns a pair of the primal and tangent components of the supplied `dx`, with
-  respect to the supplied `tag`. See the docs for [[primal-part]]
-  and [[tangent-part]] for more details.
-
-  [[primal-tangent-pair]] is equivalent to
-
-  `[([[primal-part]] dx tag) ([[tangent-part]] dx tag)]`
-
-  but slightly more efficient if you need both."
-  ([dx] (primal-tangent-pair dx (max-order-tag dx)))
-  ([dx tag]
-   (if-not (differential? dx)
-     [dx 0]
-     (let [[tangent-terms primal-terms]
-           (us/separatev #(tag-in-term? % tag)
-                         (bare-terms dx))]
-       [(terms->differential primal-terms)
-        (terms->differential tangent-terms)]))))
-
-(defn finite-term
-  "Returns the term of the supplied [[Differential]] `dx` that has no tags
-  attached to it, `0` otherwise.
-
-  [[Differential]] instances with many can be decomposed many times
-  into [[primal-part]] and [[tangent-part]]. Repeated calls
-  to [[primal-part]] (with different tags!) will eventually yield a
-  non-[[Differential]] value. If you know you want this, [[finite-term]] will
-  get you there in one shot.
-
-  NOTE that this will only work with a well-formed [[Differential]], i.e., an
-  instance with all terms sorted by their list of tags."
-  [dx]
-  (if (differential? dx)
-    (let [[head] (bare-terms dx)
-          ts     (tags head)]
-      (if (empty? ts)
-        (coefficient head)
-        0))
-    dx))
+   (some #{tag} *active-tags*)))
+
+(defn inner-tag
+  "Given any number of `tags`, returns the tag most recently bound
+  via [[with-active-tag]] (i.e., the tag connected with the _innermost_ call
+  to [[with-active-tag]]).
+
+  If none of the tags are bound, returns `(apply max tags)`."
+  [& tags]
+  (or (some (apply hash-set tags)
+            *active-tags*)
+      (apply max tags)))
+
+(defn tag+perturbation
```

Comment:
could you just return the dual number? it has the tag inside it already?

Reply:
YES that is true... I added this on my scratch branch when I still had `Differential` and it was a touch more expensive to get the tag out, so I wanted to save the effort.

Can we look at this together when I massage this to be compatible with tapes too?